### PR TITLE
Add new config option allowVisibleSource

### PR DESCRIPTION
**Pull request by roman-yatsenko**
Add new config option allowVisibleSource to the project. The default value for this option is true. The false value will block Edit and Both editor modes for users that have no rights to edit a note. 

Pull request adds:

- new config option allowVisibleSource with default value true;
- new environment variable CMD_VISIBLE_SOURCE;
- disable Edit and Both modes for users without edit note permission;
- redirect form Edit and Both modes to View mode by direct URL with mode parameter.

### CODE REVIEW

File: public/js/index.js [hunk 1]
```diff
@@ -499,6 +505,71 @@ $(window).on('error', function () {
...
+        blockSourceView = false
+      } else {
+        blockSourceView = true
+        disableControls()
+      }
+      break
+  }
+}
+
+function disableControls () {
```
Comment (a60814billy) [inline public/js/index.js:550]: Using more precise function name here.
Maybe can change `disableControls`to `disableModeChangeControls` and 
 `enableControls` to `enableModeChangeControls`

File: public/js/index.js [hunk 2]
```diff
@@ -499,6 +505,71 @@ $(window).on('error', function () {
   // setNeedRefresh();
 })
 
+function checkParametr (isLogin, permission) {
```
Comment (a60814billy) [inline public/js/index.js:508]:
```suggestion
function checkParameter (isLogin, permission) {
```

File: public/js/index.js [hunk 3]
```diff
@@ -499,6 +505,71 @@ $(window).on('error', function () {
...
+  if (typeof isLogin !== 'boolean' || !permission) {
+    throw new Error('one or more parametr is incorrect')
+  } else return allowVisibleSource(isLogin, permission)
+}
+
+function replaceUrl (url) {
```
Comment (a60814billy) [inline public/js/index.js:514]: `replaceUrl` -> `replaceUrlToViewMode`

File: public/js/index.js [hunk 4]
```diff
@@ -499,6 +505,71 @@ $(window).on('error', function () {
...
+  })
+}
+
+function enableControls () {
+  ui.toolbar.edit.removeAttr('disabled')
+  ui.toolbar.both.removeAttr('disabled')
+}
+
+function userIsLogin (userPersonalInfo) {
+  if (Object.prototype.hasOwnProperty.call(userPersonalInfo, 'login')) {
```
Comment (a60814billy) [inline public/js/index.js:565]:
```js
function userIsLogin (userPersonalInfo) {
  return !!userPersonalInfo && !!userPersonalInfo.login
}
```

File: public/js/index.js [hunk 5]
```diff
@@ -1567,10 +1638,14 @@ function importFromUrl (url) {
 
 // mode
 ui.toolbar.mode.click(function () {
```
Comment (a60814billy) [inline public/js/index.js:1640]: introduce a function `isAllowUserChangeMode`
```js
function isAllowUserChangeMode () {
  const isOwner = personalInfo.userid && window.owner && personalInfo.userid === window.owner
  return isOwner || !blockSourceView
}
```

and mode select logic would be

```js
if (isAllowUserChangeMode()) {
    return toggleMode()
}
```

**Review by a60814billy — CHANGES_REQUESTED** [review]
Hi @roman-yatsenko, thanks for your contribution, It's a good feature.
But we need some refactoring to meet project coding conventions before merge.

**Comment by a60814billy** [comment]
By the way, please sign DCO

File: lib/config/default.js [hunk 6]
```diff
@@ -33,6 +33,7 @@ module.exports = {
   allowAnonymousEdits: true,
   allowAnonymousViews: true,
   allowFreeURL: false,
+  allowVisibleSource: true,
```
[the remark this answers is not in the document]
Comment (roman-yatsenko) [reply lib/config/default.js]: With the default value `true` user can see note source (edit and both modes are enabled for view to user without edit rights). The true value is an usual CodiMD behavior for now.
P.S. But we've refactored as you wish :-)

**Comment by roman-yatsenko** [comment]
> By the way, please sign DCO

We have a git problems with DCO commit signs (Can't sign done commits). Can we accept the pull request without DCO? Or we can create a new pull request with one signed commit with all code. 
What we need to do?

**Comment by a60814billy** [comment]
Hi, @roman-yatsenko 
Due to AGPL licesne, we can't accept any code without DCO sign. Please open a new pull request to do it!

**Review by JamesHow1ett — APPROVED** [review]

**Comment by roman-yatsenko** [comment]
Recreated at https://github.com/hackmdio/codimd/pull/1621

**Pull request closed without merging**